Support input method and text input

dcz-purism opened the pull request:
This change introduces basic but useful support for `zwp-text-input-v3` and `zwp-input-method-v2` protocols.

The compositor becomes a relay between the input method program and the user-facing text input application, passing messages between them, and gracefully handling connections and disconnections.

The text-input protocol is implemented fully. The basic and least controversial pieces to make input-method functionality complete are implemented as well. Remaining are the grab and popup functionality. What's currently implemented is enough to support virtual keyboards, but probably not enough for e.g. CJK input methods.

The structure doing most of the work is the `rootston/text_input.h::roots_input_method_relay`. One is attached to each seat, and each manages multiple text-input connections and zero or one input-method connections.

Testing possible using virtboard's [input_method branch](https://source.puri.sm/Librem5/virtboard/tree/input_method) (at scaling factor x1) and [gtk-3.24](https://gitlab.gnome.org/GNOME/gtk/tree/gtk-3-24).

To test with system-wide GTK:

```
$WLROOTS/rootston/rootston &
$VIRTBOARD/virtboard &
yad --entry &
```

With locally-installed GTK, the `yad` line can look like this:

```
LD_LIBRAY_PATH=$GTK/lib yad --entry
```

or, in the unlikely case of issues:

```
GDK_BACKEND=wayland GTK_IM_MODULE=wayland GTK_IM_MODULE_FILE=$GTK/lib/gtk-3.0/3.0.0/immodules.cache yad --entry
```

dcz-purism commented:
Script for easier testing:

```
#!/bin/bash
set -e
PREFIX=`pwd`/local
BASE=`pwd`
export PATH=$PREFIX/bin/:$PATH
export PKG_CONFIG_PATH=$PREFIX/share/pkgconfig/:$PREFIX/lib/pkgconfig/

sudo apt-get -y update
sudo apt-get -y install wget git yad build-essential

# wayland 1.16 not in Debian
sudo apt-get -y install libffi-dev libxml2-dev automake libexpat1-dev
wget https://wayland.freedesktop.org/releases/wayland-1.16.0.tar.xz
tar -xf wayland-1.16.0.tar.xz
cd wayland-1.16.0
aclocal
automake
./configure --disable-documentation --prefix=$PREFIX
make -j4
make install

cd $BASE

# build virtboard
sudo apt-get -y install debhelper meson pkg-config libpixman-1-dev libpng-dev libxcb1-dev libxcb-xkb-dev libxkbcommon-dev libcairo2-dev libwayland-dev wayland-protocols
git clone https://source.puri.sm/Librem5/virtboard.git -b input_method
export LC_ALL=C.UTF-8
meson virtboard virtboard_build
ninja -C virtboard_build

# build rootston
sudo apt-get -y install debhelper libcap-dev libdrm-dev libegl1-mesa-dev libgbm-dev libgles2-mesa-dev libinput-dev libpixman-1-dev libsystemd-dev libwayland-dev libxcb1-dev libxcb-composite0-dev libxcb-icccm4-dev libxcb-image0-dev libxcb-render0-dev libxcb-xfixes0-dev libxkbcommon-dev meson pkg-config wayland-protocols build-essential
git clone https://github.com/dcz-purism/wlroots.git -b input
meson wlroots wlroots_build
ninja -C wlroots_build

# build GTK
sudo apt-get -y install gtk-doc-tools gobject-introspection libatk1.0-dev libgirepository1.0-dev libepoxy-dev libpango1.0-dev libgdk-pixbuf2.0-dev
git clone --depth 1 https://gitlab.gnome.org/GNOME/gtk.git -b gtk-3-24
cd gtk
./autogen.sh --disable-x11-backend --enable-wayland-backend --disable-documentation --prefix=$PREFIX
make -j4
make install

exit 0

# run a demo - better play with all the programs yourself

cd $BASE
./wlroots_build/rootston/rootston &
./virtboard_build/virtboard &
yad &
LD_LIBRARY_PATH=$PREFIX/lib yad --entry &
```

agx commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+	context->pending.preedit.cursor_end = cursor_end;
+	free(context->pending.preedit.text);
+	context->pending.preedit.text = strdup(text); // TODO: clean up
+}
```
This leaks on destroy.

agx commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
+	free(context->pending.commit_text);
+	context->pending.commit_text = strdup(text); // TODO: clean up
```
This leaks on destroy.

agx commented on examples/meson.build:
```diff
@@ -90,6 +90,14 @@ examples = {
+	'text-input': {
+	    'src': 'text-input.c',
+	    'dep': [wayland_cursor, wayland_client, wlr_protos, wlroots],
+    },
```
You're adding both in the same commi alhough one should be in text-input and the other one in input-method.

dcz-purism replied:
Should be fixed now.

SirCmpwn commented on include/wlr/types/wlr_input_method.h:
```diff
@@ -0,0 +1,73 @@
+#include <stdlib.h>
+#include <wayland-server.h>
+
+
```
Extra newline here and on line 3

SirCmpwn commented on include/wlr/types/wlr_input_method.h:
```diff
@@ -0,0 +1,73 @@
+	struct wl_global *input_method_manager;
+	struct wl_resource *resource; // FIXME: turn into a list, each entry per-client
+
+	struct wlr_input_method *input_method; // FIXME: turn into a list, each entry per-seat
```
Can you fix this now?

dcz-purism replied:
The first FIXME pertains to a list of all input methods - since input methods are meant to take over inputs using grabs, does it make sense to allow multiple ones?

SirCmpwn commented on types/wlr_text_input.c:
```diff
@@ -0,0 +1,285 @@
+		void *data) {
+	struct wlr_text_input *text_input = wl_container_of(listener, text_input,
+		seat_destroy_listener);
+//	wlr_text_input_destroy(text_input); TODO: what to do when the seat goes under?
```
become inert, probably

Shugyousha commented on types/wlr_text_input.c:
```diff
@@ -0,0 +1,285 @@
+static void text_input_enable(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_text_input *text_input = text_input_from_resource(resource);
+	struct wlr_text_input_properties defaults = {0};
```
This could be static.

Shugyousha commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+	.preedit_string = preedit_string,
+	.delete_surrounding_text = delete_surrounding_text,
+	.get_input_popup_surface = noop,
+	.grab_keyboard = noop,
```
I assume the noops here are still on the TODO list, correct?

Shugyousha commented:
[2018-08-31 07:17] dcz-purism <notifications@github.com>
> Script for easier testing:
> 
> [...]

I tested the basic virtual keyboard functionality using the instructions
you provided and can confirm that it works!

It would definitely be good to be able to do some testing with more
complex input methods, i. e. ones that show word suggestions
for example. Is the plan to add something like that to
wlroots/examples/input-method.c?

emersion commented on examples/input-method.c:
```diff
@@ -0,0 +1,214 @@
+
+static void handle_unavailable(void *data,
+		struct zwp_input_method_v2 *zwp_input_method_v2) {
+	printf("IM disappeared");
```
Missing `\n`

emersion commented on examples/text-input.c:
```diff
@@ -0,0 +1,197 @@
+#include "text-input-unstable-v3-client-protocol.h"
+#include "xdg-shell-client-protocol.h"
+
+#include <linux/input-event-codes.h>
```
Add in FreeBSD too. You can find examples of how to do this by grepping other examples.

emersion commented on examples/input-method.c:
```diff
@@ -0,0 +1,214 @@
+#include "input-method-unstable-v2-client-protocol.h"
+#include "xdg-shell-client-protocol.h"
+
+#include <linux/input-event-codes.h>
```
FreeBSD compat

emersion commented on include/wlr/types/wlr_input_method.h:
```diff
@@ -0,0 +1,73 @@
+#ifndef WLR_TYPES_WLR_INPUT_METHOD_H
```
This file should have a `_v2` suffix

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef WLR_TYPES_WLR_TEXT_INPUT_H
```
This file should have a `_v3` suffix (alongside with all structs).

Ah, and you need to add the `WLR_USE_UNSTABLE` stuff too at the beginning of all new header files.

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+};
+
+struct wlr_text_input_manager {
+	struct wl_global *wl_global;
```
Should be named `global`

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+
+	struct {
+		struct wl_signal text_input; // (struct wlr_text_input*)
+	} events;
```
Should have a `destroy` event

emersion replied:
I mean a destroy event on the manager. The manager gets destroyed with the display. If the compositor has objects that depend on it it needs to destroy them with the manager (it can't destroy them with the display because the destroy listener of the manager may be triggered before the destroy listener of the objects, this has caused issues in the past).

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+
+struct wlr_text_input_manager {
+	struct wl_global *wl_global;
+	struct wl_list text_inputs; // wlr_text_input::resource::link
```
Hmm, you mean `wlr_text_input::link`?

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_surface.h>
+
+struct wlr_text_input_properties {
```
We generally use `state` instead of `properties`

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+		uint32_t anchor;
+	} surrounding;
+
+	uint32_t text_change_cause;
```
Can this be an `enum`?

emersion commented on include/wlr/types/wlr_text_input.h:
```diff
@@ -0,0 +1,77 @@
+	struct wlr_surface *focused_surface;
+	struct wlr_text_input_properties pending;
+	struct wlr_text_input_properties current;
+	uint32_t current_serial; // next to send
```
If it's the next serial, maybe we should rename to `next_serial`?

emersion commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+static void commit(struct wl_client *client, struct wl_resource *resource,
+		uint32_t serial) {
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
```
Should add a safe wrapper `input_method_from_resource` for this

emersion commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+	context->current_serial = serial;
+	struct wlr_input_method_state default_state = {0};
+	context->pending = default_state;
+	wlr_signal_emit_safe(&context->events.commit, (void*)context);
```
No need to cast to `void *`

emersion commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+
+static void input_method_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_input_method *context = input_method_from_resource(resource);
+	if (!context) {
```
`input_method_destroy` should check for NULL

emersion replied:
No, but that's how inert resources work. See the last part of `CONTRIBUTING.md`.

dcz-purism replied:
Thanks, I didn't realize this guide was there.

emersion commented on types/wlr_input_method.c:
```diff
@@ -0,0 +1,212 @@
+
+static void manager_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	// FIXME
```
This should call `wl_resource_destroy`

emersion commented on types/wlr_text_input.c:
```diff
@@ -0,0 +1,285 @@
+	wl_list_remove(&text_input->link);
+	// no need to destroy the resource if client closed, prompting the server to do it
+	if (text_input->resource) {
+		wl_resource_destroy(text_input->resource);
```
We should never do this. Instead, the client should always send a request to destroy the resource.

emersion commented on types/wlr_text_input.c:
```diff
@@ -0,0 +1,285 @@
+
+static void text_input_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	text_input_resource_destroy(resource);
```
This should just call `wl_resource_destroy`

dcz-purism commented:
That's a very timely bump, I'm working on addressing these issues ATM. I should have a new update out within a couple of days.

SirCmpwn commented:
Nice.

dcz-purism commented:
The general issues should be taken care of, examples work in progress.

dcz-purism commented on protocol/meson.build:
```diff
@@ -37,7 +39,9 @@ client_protocols = [
+	'input-method-unstable-v2.xml',
 	'screenshooter.xml',
+	'text-input-unstable-v3.xml',
 	'wlr-export-dmabuf-unstable-v1.xml',
```
Should be fixed now.

dcz-purism commented:
Update: the input-metod example is working nice now.

dcz-purism commented:
Pushed new update - the text input example is much nicer now. I didn't know how to make text display on an OpenGLES surface easily, so skipped that part in favor of console output.

SirCmpwn commented on examples/input-method.c:
```diff
@@ -0,0 +1,393 @@
+#include <stdlib.h>
+#include <string.h>
+#include <sys/epoll.h>
+#include <sys/timerfd.h>
```
They're not standard but are supported by the OSes we support.

Shugyousha commented:
[2018-09-28 18:59] dcz-purism <notifications@github.com>
>
> Update: the input-metod example is working nice now.

Thanks for the update!

I ran input-method with rootston and then tried to run yad to test some
input. I used the following command to start yad:

WAYLAND_DEBUG=1 GDK_BACKEND=wayland GTK_IM_MODULE=wayland yad --entry

I got the following output from input-method and then yad crashed with
a segmentation fault:

Timer armed


Handle serial 1
Now active
Surrounding text:
Cursor after 0:
Update 0
Timer armed
Handle serial 2
Update 1
Handle serial 3
Update 2
Handle serial 4
Update 2
Handle serial 5
Update 2
Handle serial 6
Surrounding text: _Commit_
Cursor after 8: _Commit_
Update 2
Timer armed
Handle serial 7
Update 3
Handle serial 8
Update 3
Handle serial 9
Surrounding text: _Commit__CommitNoPreed_
Cursor after 23: _Commit__CommitNoPreed_
Update 3
Handle serial 10
Now inactive
Update 4

Is this expected? Is there another way to test it?

I am running gtk3 3.24 which I think should support the necessary version
of the Wayland text-input protocol.

SirCmpwn commented:
Does virtboard support both of these protocols yet?

Shugyousha commented:
[2018-09-30 09:34] dcz-purism <notifications@github.com>
> Pushed new update - the text input example is much nicer now. I didn't
> know how to make text display on an OpenGLES surface easily, so
> skipped that part in favor of console output.

A few month ago I struggled with the same issue. I copied a lot of
code from some other repo and got some text to show using wl_shm and
cairo. It's not OpenGLES and it's very crappy *but* it shows (still
static) text :P if you want to have a look, you can find it here:

https://github.com/Shugyousha/wlroots/commits/cairowayland

I ran the new text-input together with the new input-method and there
were no crashes. Since I cannot find a way to get my keyboard input to
show I assume that it's working as expected.

SirCmpwn commented:
I'm in favor of just using the console and skipping text. The only way to do text in GLES is with an external text rendering dependency like pango/harfbuzz.

SirCmpwn commented:
Well, I imagine your eventual plan is to provide an OSK which uses text-input to provide suggestions and so on.

dcz-purism commented:
Yes, that's our long term goal. Virtboard is just a stopgap measure.

Btw, the master branch has all necessary changes now: https://source.puri.sm/Librem5/virtboard/

Shugyousha commented:
[2018-09-30 10:26] dcz-purism <notifications@github.com>
>
> @Shugyousha GTK will crash on delete_surrounding_text, you need to use
> the git version of the 3.24 immodule (sorry for not mentioning).

No problem. That's what I tried first but I couldn't compile it because it
insists on an older version of automake than Arch currently provides... I
may give it another shake later.

Shugyousha commented:
[2018-09-30 09:57] Drew DeVault <notifications@github.com>
>
> I'm in favor of just using the console and skipping text. The only way
> to do text in GLES is with an external text rendering dependency like
> pango/harfbuzz.

Yes, I think that makes sense. Testing/demonstrating the
input-method/text-input interactions should be well possible with just
printing to stdout.

dcz-purism commented:
Another bump - I removed some duplicated/convoluted state and am sending enter/leave events when the input method connects/disconnects.

I think this is close to the final state now (modulo unimplemented pieces).

dcz-purism commented on types/wlr_text_input_v3.c:
```diff
@@ -0,0 +1,333 @@
+		// GTK currently assumes that input method focus follows keyboard focus
+		//wl_resource_post_error(text_input->resource, 0, "Text input was not"
+			//"entered, and cannot be enabled\n");
+		wlr_log(WLR_ERROR, "Received mistaken request to enable text-input");
```
This is the last issue I know about, and I'm working on the GTK side now.

dcz-purism commented:
The most recent push solves the last issue and only the last issue: https://gitlab.gnome.org/GNOME/gtk/merge_requests/369

dcz-purism commented:
The current status:

- text-input (text consumer) is covered 100%, although there's no one to consume cursor rectangle
- input-method (text producer) is missing some pieces, but covers 100% of the core, that is, it supports everything that an input method needs to communicate with an application to the fullest extent

The missing pieces are:

- a keyboard grab interface for input methods (need to be reworked in zwp-input-method-v2 proposal)
- floating pop-ups

The missing features are relevant for CJK languages, but the done ones are useful for Latin input methods. I would appreciate if this branch got merged separately, and then I could work on CJK without worrying about rebasing heaps of code :)

SirCmpwn commented on types/wlr_input_method_v2.c:
```diff
@@ -0,0 +1,287 @@
+	.preedit_string = im_preedit_string,
+	.delete_surrounding_text = im_delete_surrounding_text,
+	.get_input_popup_surface = noop, // TODO: implement
+	.grab_keyboard = noop, // TODO: implement
```
I'd rather have two real placeholder functions than one noop function

SirCmpwn commented:
This looks pretty good. I'd like to see the minor comments addressed and get a second opinion, then I'm in favor of merge.

dcz-purism commented on examples/text-input.c:
```diff
@@ -0,0 +1,385 @@
+	// glyphs
+	for (unsigned i = 0; i < utf8_strlen(buffer); i++) {
+		printf(" ");
+	}
```
That's an important note, I'm going to put it in the usage string.

dcz-purism commented:
I addressed the issues and updated the input-method protocol to the new patch version to get rid of many mistakes in description texts. This implied one code change: `preedit_string` calls became `set_preedit_string`.

dcz-purism commented:
Fixed usage strings.

dcz-purism commented:
Did the build actually take place? There's not much in the logs...

emersion commented:
Restarted the build, seems to work better: https://builds.sr.ht/~emersion/job/8183

Shugyousha commented on examples/text-input.c:
```diff
@@ -0,0 +1,393 @@
+		cursor_mark[i] = '.';
+	}
+	if (current.preedit.cursor_begin == -1
+			&& current.preedit.cursor_end == -1) {
```
cursor_mark leaks here and at the other returns (where we don't goto end). Found by clang https://builds.sr.ht/~emersion/job/8183#task-clang-334 .

dcz-purism replied:
Fixed

dcz-purism commented:
Bump

SirCmpwn commented:
Well, I like it. Thanks!

dcz-purism commented:
Thank you!

dcz-purism commented:
@Edootjuh I can't reproduce this, how are you compiling it?

agx commented:
@SirCmpwn thanks for fixing this up so quickly.

SirCmpwn commented:
@dcz-purism I pushed a fix
